### ISSUE TITLE
 Import some data to existing PI points from a CSV File

### ISSUE THREAD

**Issue by pthivierge** [osisoft/PI-AF-SDK-Clues#1]
Description from Lonnie:
Just a simple csv file with each row as a timestamp + values, and the columns (header), the point names (this could be any number of points).

**Comment by pthivierge** [comment]
This use case is very interesting, there is a small difficulty where the data format could be different from one use case to another.  So we'd need to either implement one way, that we believe can be common for most of the use cases, or find a way to provide the utility with a template that will define how the data is stored in the data file(s).

**Comment by LonnieBowling** [comment]
To keep it simple, I think just have a fixed file format:

TimeStamp, Tag1, Tag2, Tag3, ...
DateTime, Value1, Value2, Value3 ...
DateTime, Value1, Value2, Value3 ...
...

We can use the point information in PI to figure out the type (numeric, digtal, string, etc...) and try to do the conversion to the proper type before writing to PI. If the conversion fails the operation stops. It may be nice to be able to run the command without actually writing the values to PI just to make sure all the conversions are OK.

**Comment by pthivierge** [comment]
This use case looks good Lonnie.

I would probably also look for: 
TimeStamp, Tag1, Value1
TimeStamp, Tag2, Value2
...
as sometimes values are not at the same timestamp.

I'll give it a shot in the next coming days.

**Comment by LonnieBowling** [comment]
Good point about timestamps, how are you going to handle the application being able to recognize the two different file formats?

**Comment by pthivierge** [comment]
I'll start with the first scenario, and see how this can be generalized / refactored... I have to put my hands in the dirt to see how this can be, perhaps the easiest will be to implement the data parsing part as an Interface, so we can expand by writing creating new classes based on this interface. You'd have to tell about the file format at the command line. If you have suggestions they are welcome.

**Comment by pthivierge** [comment]
Hello Lonnie,  I could not yet find the time to implement this feature, however I did some other work in the dev branch where I needed to use load data from a text file. See [Library.DataHandling.FileLoaded](https://github.com/osisoft/PI-AF-SDK-Clues/blob/dev/Core/Library/DataHandling/FileLoader.cs) class.  This class can read data with headers, that are copy-pasted from Excel (Tab character as a delimiter) to a text file.  Columns that are unknown are ignored. it could be easily transformed to read csv instead.  And this could be a good start I believe because it builds a data table and that can be easily manipulated.  With some naming conventions for column names, we could probably handle all the cases we talked about in this conversation.  Have a good Christmas time ;)
